core/ledger/kvledger/kv_ledger_test.go

```diff
@@ -1118,6 +1119,144 @@ func TestCollectionConfigHistoryRetriever(t *testing.T) {
 	})
 }
 
+func TestCommitNotifications(t *testing.T) {
+	var lgr *kvLedger
+	var doneChannel chan struct{}
+	var dataChannel <-chan *ledger.CommitNotification
+
+	setup := func() {
+		var err error
+		lgr = &kvLedger{}
+		doneChannel = make(chan struct{})
+		dataChannel, err = lgr.CommitNotificationsChannel(doneChannel)
+		require.NoError(t, err)
+	}
+
+	t.Run("only first txid is included in notification", func(t *testing.T) {
+		setup()
+		lgr.sendCommitNotification(1, []*validation.TxStatInfo{
+			{
+				TxIDFromChannelHeader: "txid_1",
+				ValidationCode:        peer.TxValidationCode_BAD_RWSET,
+			},
+			{
+				TxIDFromChannelHeader: "txid_1",
+				ValidationCode:        peer.TxValidationCode_DUPLICATE_TXID,
+			},
+			{
+				TxIDFromChannelHeader: "txid_2",
+				ValidationCode:        peer.TxValidationCode_VALID,
+			},
+		})
+
+		commitNotification := <-dataChannel
+		require.Equal(t,
+			&ledger.CommitNotification{
+				BlockNumber: 1,
+				TxIDValidationCodes: map[string]peer.TxValidationCode{
+					"txid_1": peer.TxValidationCode_BAD_RWSET,
+					"txid_2": peer.TxValidationCode_VALID,
+				},
+			},
+			commitNotification,
+		)
+	})
+
+	t.Run("empty txids are not included in notification", func(t *testing.T) {
+		setup()
+		lgr.sendCommitNotification(1, []*validation.TxStatInfo{
+			{
+				TxIDFromChannelHeader: "",
+				ValidationCode:        peer.TxValidationCode_BAD_RWSET,
+			},
+			{
+				TxIDFromChannelHeader: "",
+				ValidationCode:        peer.TxValidationCode_DUPLICATE_TXID,
+			},
+		})
+
+		commitNotification := <-dataChannel
+		require.Equal(t,
+			&ledger.CommitNotification{
+				BlockNumber:         1,
+				TxIDValidationCodes: map[string]peer.TxValidationCode{},
+			},
+			commitNotification,
+		)
+	})
+
+	t.Run("second time calling CommitNotificationsChannel returns error", func(t *testing.T) {
+		setup()
+		_, err := lgr.CommitNotificationsChannel(make(chan struct{}))
+		require.EqualError(t, err, "only one commit notifications channel is allowed at a time")
+	})
+
+	t.Run("closing done channel closes the data channel on next commit", func(t *testing.T) {
+		setup()
+		lgr.sendCommitNotification(1, []*validation.TxStatInfo{})
+		_, ok := <-dataChannel
+		require.True(t, ok)
+
+		close(doneChannel)
+		lgr.sendCommitNotification(2, []*validation.TxStatInfo{})
+
+		_, ok = <-dataChannel
+		require.False(t, ok)
+	})
+}
+
+func TestCommitNotificationsOnBlockCommit(t *testing.T) {
+	conf, cleanup := testConfig(t)
+	defer cleanup()
+	provider := testutilNewProvider(conf, t, &mock.DeployedChaincodeInfoProvider{})
+	defer provider.Close()
+
+	bg, gb := testutil.NewBlockGenerator(t, "testLedger", false)
+	protoutil.BlockHeaderHash(gb.Header)
```

Comment:
What's this for? It's calculating the hash on the block bytes but throwing it away.

Reply:
done